### PR TITLE
(PUP-10677) gem provider uses system ruby

### PR DESCRIPTION
**Pull request by gimmyxd**
Because `INSTALL_DIR/puppet/bin` is prepended to PATH,
gem provider always uses Puppet's Ruby.
This commit changes the gem provider to execute commands
in a specific environment that has a PATH which does not
contain `puppet/bin`.

It also updates the puppet_gem provider to use
the full path for the Puppet's gem executable

### CODE REVIEW

**Comment by gimmyxd** [comment]
this should replace the following:
 - https://github.com/puppetlabs/puppet-agent/pull/1973
 - https://github.com/puppetlabs/puppet/pull/8335

**Comment by joshcooper** [comment]
Needs rebase on upstream/main to get the rubocop fix in c3a37a92d5

File: lib/puppet/provider/package/gem.rb [hunk 1]
```diff
@@ -46,6 +50,18 @@ def self.provider_command
     is_optional
   end
 
+  # Having puppet/bin in PATH makes gem provider to use puppet/bin/gem
+  # This is an utiliy methods that reads the PATH and returns a string
+  # that contains the content of PATH but without puppet/bin dir.
+  # This is used to pass a custom PATH as a specific environment for executed commands
+  def self.windows_path_without_puppet_bin
+    @path ||= Puppet::Util.get_env('PATH').split(File::PATH_SEPARATOR)
+                                          .reject { |dir| dir =~ /puppet\\bin/ }
```
Comment (joshcooper) [inline lib/puppet/provider/package/gem.rb:59]: Does puppet's bin directory always contain backslashes in the `PATH` environment variable? Maybe use `Pathname#cleanpath` to normalize?

```
irb(main):002:0> Pathname.new("C:\\Puppet").cleanpath
=> #<Pathname:C:/Puppet>
```

Should we only reject entries that end with `/puppet\\bin$/`?

Comment (gimmyxd) [reply lib/puppet/provider/package/gem.rb:59]: > Does puppet's bin directory always contain backslashes in the PATH environment variable?

I believe so, since this is controlled in environment.rb(https://github.com/puppetlabs/puppet-agent/blob/master/resources/files/windows/environment.bat#L13) and we also use backslashes in daemon:(https://github.com/puppetlabs/puppet/blob/main/ext/windows/service/daemon.rb#L203-L206)

> Should we only reject entries that end with /puppet\\bin$/?

Yes

File: spec/unit/provider/package/puppet_gem_spec.rb [hunk 2]
```diff
@@ -15,19 +15,20 @@
...
   if Puppet::Util::Platform.windows?
-    let(:provider_gem_cmd) { 'gem' }
+    let(:provider_gem_cmd) { 'C:\Program Files\Puppet Labs\Puppet\puppet\bin\gem.bat' }
   else
     let(:provider_gem_cmd) { '/opt/puppetlabs/puppet/bin/gem' }
   end
 
   custom_environment = {"HOME"=>ENV["HOME"]}
-  custom_environment['PKG_CONFIG_PATH'] = '/opt/puppetlabs/puppet/lib/pkgconfig' unless Puppet::Util::Platform.windows?
+  custom_environment['PKG_CONFIG_PATH'] = '/opt/puppetlabs/puppet/lib/pkgconfig'
```
Comment (GabrielNagy) [inline spec/unit/provider/package/puppet_gem_spec.rb:24]: Is this change needed?

Comment (gimmyxd) [reply spec/unit/provider/package/puppet_gem_spec.rb:24]: Yes, unit tests should not be platform dependent.

File: spec/unit/provider/package/gem_spec.rb [hunk 3]
```diff
@@ -36,6 +40,32 @@
...
+        let(:expected_path) do
+          "C:\\Program Files\\Puppet Labs\\Puppet\\bin;C:\\Ruby26-x64\\bin;C:\\Windows\\system32\\bin"
+        end
+        before do
+          allow(Puppet::Util::Platform).to receive(:windows?).and_return(true)
+          allow(Puppet::Util).to receive(:get_env)
+          allow(Puppet::Util).to receive(:get_env).with('PATH').and_return(path)
+          allow(described_class).to receive(:validate_command).with(provider_gem_cmd)
+          stub_const('::File::PATH_SEPARATOR', ';')
+        end
```
Comment (GabrielNagy) [inline spec/unit/provider/package/gem_spec.rb:56]: nit: would add some empty lines between let/let/before for consistency; also you have a whitespace in the first `let (:path)`

**Review by GabrielNagy — APPROVED** [review]

File: spec/unit/provider/package/gem_spec.rb [hunk 4]
```diff
@@ -36,6 +40,34 @@
...
+        let(:expected_path) do
+          "C:\\Program Files\\Puppet Labs\\Puppet\\bin;C:\\Ruby26-x64\\bin;C:\\Windows\\system32\\bin"
+        end
+
+        before do
+          allow(Puppet::Util::Platform).to receive(:windows?).and_return(true)
+          allow(Puppet::Util).to receive(:get_env)
+          allow(Puppet::Util).to receive(:get_env).with('PATH').and_return(path)
+          allow(described_class).to receive(:validate_command).with(provider_gem_cmd)
+          stub_const('::File::PATH_SEPARATOR', ';')
```
Comment (joshcooper) [inline spec/unit/provider/package/gem_spec.rb:57]: We'll need to restore this when done. Maybe easier to just confine the `on windows` context to `if: Puppet::Util::Platform.windows?` instead of trying to run it everywhere.

File: spec/unit/provider/package/gem_spec.rb [hunk 5]
```diff
@@ -36,6 +40,34 @@
...
+        end
+
+        let(:expected_path) do
+          "C:\\Program Files\\Puppet Labs\\Puppet\\bin;C:\\Ruby26-x64\\bin;C:\\Windows\\system32\\bin"
+        end
+
+        before do
+          allow(Puppet::Util::Platform).to receive(:windows?).and_return(true)
+          allow(Puppet::Util).to receive(:get_env)
+          allow(Puppet::Util).to receive(:get_env).with('PATH').and_return(path)
```
Comment (joshcooper) [inline spec/unit/provider/package/gem_spec.rb:55]: puppet's spec_helper restores the environment across tests (https://github.com/puppetlabs/puppet/blob/b3fe419f5170d8726f493d707b3751ccbf7c1bbd/lib/puppet/test/test_helper.rb#L194-L201) so you could avoid stubbing and do:

```
ENV['PATH'] = "#{path};#{ENV['PATH']}"
```

But then we're mucking with the running ruby process' PATH...

File: spec/unit/provider/package/gem_spec.rb (hunk 4 above)
Comment (gimmyxd) [reply spec/unit/provider/package/gem_spec.rb:57]: I don't really like the idea of having unit tests behaving differently depending on the platform they are run.

[stub_const](https://relishapp.com/rspec/rspec-mocks/v/2-11/docs/stubbing-constants) will restore to the original value after the example is run so no need to take care of that.

```ruby

  context 'when stubbing' do
    before do
      stub_const('::File::PATH_SEPARATOR', ';')
    end
    it 'is ;' do
      puts "::File::PATH_SEPARATOR => #{::File::PATH_SEPARATOR}"
      expect(::File::PATH_SEPARATOR).to eql(';')
    end
  end

  context 'after stubbing' do
    it do
      puts "::File::PATH_SEPARATOR => #{::File::PATH_SEPARATOR}"
      expect(::File::PATH_SEPARATOR).not_to eql(';')
    end
  end
end
```

```
description
  when stubbing
::File::PATH_SEPARATOR => ;
    is ;
  after stubbing
::File::PATH_SEPARATOR => :
    is expected not to eql ";"
```

File: spec/unit/provider/package/gem_spec.rb (hunk 5 above)
Comment (gimmyxd) [reply spec/unit/provider/package/gem_spec.rb:55]: I think is better contained if we mock only the part that we are interested in instead of the `PATH` env. I wouldn't expect that mocking `get_env('PATH')` to interfere with other parts of puppet and make this test unreliable.

**Review by ciprianbadescu — APPROVED** [review]

File: spec/unit/provider/package/gem_spec.rb (hunk 4 above)
Comment (joshcooper) [reply spec/unit/provider/package/gem_spec.rb:57]: > I don't really like the idea of having unit tests behaving differently depending on the platform they are run.

True, but the behavior *is* different on windows vs others, so testing the windows behavior on non-windows platforms isn't super useful. You'd have to also stub `::File::ALT_SEPARATOR` and `::File::SEPARATOR` in order to more closely mimic Windows (otherwise `C:\\Program Files\\Puppet Labs\\Puppet\\puppet\\bin;C:\\Program Files\\Puppet Labs\\Puppet\\bin;C:\\Ruby26-x64\\bin;C:\\Windows\\system32\\bin"` would appear to be a relative path for a single file).

Comment (joshcooper) [reply spec/unit/provider/package/gem_spec.rb:57]: > stub_const will restore to the original ...

My bad, I misread that. I thought it said `const_set`

Comment (joshcooper) [reply spec/unit/provider/package/gem_spec.rb:57]: > isn't super useful

That said I'm fine to merge as is

**Review by joshcooper — APPROVED** [review]
